### PR TITLE
upgrade to sharing 2.0, remove hierarchical re-shares

### PR DESCRIPTION
**Pull request by schiessle**
Update script to remove all hierarchical re-shares and create a flat list instead.

I will add unit test over the next days... Feel free to have a look if you see stuff which could be optimized/improved.

cc @rullzer @DeepDiver1975

### CODE REVIEW

**Comment by rullzer** [comment]
Great start!

File: apps/files_sharing/lib/migration.php [hunk 1]
```diff
@@ -20,22 +19,126 @@
...
+	/**
+	 * get all re-shares from the database
+	 *
+	 * @return array
+	 */
+	private function getAllReShares() {
+		$query = $this->connection->getQueryBuilder();
+		$query->select(['id', 'parent', 'uid_owner'])
+			->from($this->table)
+			->orWhere($query->expr()->eq('share_type', $query->createNamedParameter(\OCP\Share::SHARE_TYPE_USER)))
```
Comment (DeepDiver1975) [inline apps/files_sharing/lib/migration.php]: use  in() ??? - just an idea - no clue if that could help with query execution ...

Comment (schiesbn) [inline apps/files_sharing/lib/migration.php]: I'm also not sure if there is a performance difference... still makes probably sense to use the in operator here

**Comment by schiesbn** [comment]
While writing tests I wonder if we also want to move the password for link shares from the "share_with" column to a new column "password"? I always hated that we abuse the share_with column for the password. I think now it would be the right time to change it... What do you think @DeepDiver1975 @rullzer ?

**Comment by rullzer** [comment]
@schiesbn We can only do that once we have moved all the internal handling to that column as well... so basically after all is done we can move it.. but now there are still to many places that use the old code path...

**Comment by rullzer** [comment]
@schiesbn but i agree it has to go :P

**Comment by schiesbn** [comment]
> We can only do that once we have moved all the internal handling to that column as well... so basically after all is done we can move it.. but now there are still to many places that use the old code path...

OK, so let's do this in a second step.

**Comment by DeepDiver1975** [comment]
ci error:

````
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to OCA\Files_Sharing\Migration::__construct() must implement interface OCP\IDBConnection, none given, called in /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/apps/files_sharing/tests/migrationtest.php on line 63 and defined in /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/apps/files_sharing/lib/migration.php:37
11:03:46 Stack trace:
11:03:46 #0 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/apps/files_sharing/tests/migrationtest.php(63): OCA\Files_Sharing\Migration->__construct()
11:03:46 #1 [internal function]: MigrationTest->testAddAccept()
11:03:46 #2 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(865): ReflectionMethod->invokeArgs(Object(MigrationTest), Array)
11:03:46 #3 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(743): PHPUnit_Framework_TestCase->runTest()
11:03:46 #4 phar:///usr/local/bin/phpunit/php-invoker/Invoker.php(93): PHPUnit_Framework_TestCase->runBare()
11:03:46 #5 phar:///usr/local/bin/phpu in /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/apps/files_sharing/lib/migration.php on line 37

**Comment by schiesbn** [comment]
@DeepDiver1975 this are still the tests for the old migration class... Ignore it for now. Updated unit tests will come soon :smile:

File: apps/files_sharing/lib/migration.php [hunk 2]
```diff
@@ -20,22 +19,126 @@
...
+		foreach ($owners as $id => $owner) {
+			$query->update($this->table)
+				->set('parent', $query->createNamedParameter(null))
+				->set('uid_owner', $query->createNamedParameter($owner['owner']))
+				->set('uid_initiator', $query->createNamedParameter($owner['initiator']))
+				->where($query->expr()->eq('id', $query->createNamedParameter($id)))
+				->execute();
+		}
+
+		$this->connection->commit();
```
Comment (MorrisJobke) [inline apps/files_sharing/lib/migration.php]: What happens on commit failure? :see_no_evil: Is it still consistent and can be called a second time?

Comment (schiesbn) [inline apps/files_sharing/lib/migration.php]: yes, if the commit fails nothing was changed and you can call it a second time

**Comment by DeepDiver1975** [comment]
pgsql issue:

````
MigrationTest::testRemoveReShares
Doctrine\DBAL\Exception\SyntaxErrorException: An exception occurred while executing 'UPDATE "oc_share" SET "parent" = ?, "uid_owner" = ?, "uid_initiator" = ?, "parent" = ?, "uid_owner" = ?, "uid_initiator" = ? WHERE "id" = ?' with params [null, "owner2", "user1", null, "owner2", "user2", 346]:

SQLSTATE[42601]: Syntax error: 7 ERROR:  multiple assignments to same column "parent"

/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php:66
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:116
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:996
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/lib/private/db/connection.php:203
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:208
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/lib/private/db/querybuilder/querybuilder.php:124
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/apps/files_sharing/lib/migration.php:156
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/apps/files_sharing/lib/migration.php:57
/ssd/jenkins/workspace/core-ci-linux@6/database/pgsql/label/SLAVE/apps/files_sharing/tests/migrationtest.php:215

**Comment by DeepDiver1975** [comment]
oracel issue:

````
MigrationTest::testRemoveReShares
Doctrine\DBAL\Driver\OCI8\OCI8Exception: ORA-02289: sequence does not exist

/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Exception.php:33
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php:214
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Connection.php:116
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Connection.php:157
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1087
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/lib/private/db/connection.php:226
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/lib/private/db/adapteroci8.php:36
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/lib/private/db/connection.php:221
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/apps/files_sharing/tests/migrationtest.php:144
/ssd/jenkins/workspace/core-ci-linux@2/database/oci/label/SLAVE/apps/files_sharing/tests/migrationtest.php:214

**Comment by schiesbn** [comment]
I fixed the pgsql error, but I have no idea oci doesn't like ````$parent = $this->connection->lastInsertId($this->table);```` https://github.com/owncloud/core/pull/21660/files#diff-9617a0875321cd12aaccd88a631e5dadR144 @DeepDiver1975 any idea?

File: apps/files_sharing/tests/migrationtest.php [hunk 3]
```diff
@@ -31,51 +31,218 @@
...
+			->setParameter('item_source', '2')
+			->setParameter('item_target', '/2')
+			->setParameter('file_source', 2)
+			->setParameter('file_target', '/foo')
+			->setParameter('permissions', 31)
+			->setParameter('stime', time());
+		$this->assertSame(1,
+			$query->execute()
+		);
+		$parent = (int)$this->connection->lastInsertId('*PREFIX*' . $this->table);
```
Comment (rullzer) [inline apps/files_sharing/tests/migrationtest.php]: aaah you forgot the backticks!

**Comment by schiesbn** [comment]
now it should work... let's cross the fingers :wink:

**Comment by schiesbn** [comment]
perfect... all tests passed. Failing tests are unrelated (````cURL error 7: Failed to connect to localhost port 8083: Connection refused````)

File: apps/files_sharing/lib/migration.php [hunk 4]
```diff
@@ -20,22 +19,156 @@
...
+			->andWhere($query->expr()->in(
+				'item_type',
+				$query->createNamedParameter(
+					['file', 'folder'],
+					Connection::PARAM_STR_ARRAY
+				)
+			))
+			->andWhere($query->expr()->isNotNull('parent'));
+		$result = $query->execute();
+		$shares = $result->fetchAll();
```
Comment (DeepDiver1975) [inline apps/files_sharing/lib/migration.php]: Thinking about big installations this can be quite a lot. An iterativ approach would be better.

Comment (rullzer) [inline apps/files_sharing/lib/migration.php]: While I think even for big installations the number of reshares is still manageable. I do agree that iterative approach would be safer.

File: apps/files_sharing/lib/migration.php [hunk 5]
```diff
@@ -20,22 +19,156 @@
  *
  */
 
- namespace OCA\Files_Sharing;
+namespace OCA\Files_Sharing;
 
+use Doctrine\DBAL\Connection;
```
Comment (rullzer) [inline apps/files_sharing/lib/migration.php]: Nothing you can fix here. But it is kinda weird this needs to be included here. Breaks the abstraction

**Comment by rullzer** [comment]
Ok updated this PR.

* Queries are now done in chunks of 1000 shares
* Cache caches max 10k shares (~20MB).
* Also migrate non reshares to new format (owner/initiator).

Everything still passes.

**Comment by rullzer** [comment]
I tested locally and for me migrating approx 100k shares takes 2 minutes and 40 seconds. This is including ~500K assertions. Wich seems very reasonable.

**Comment by rullzer** [comment]
@schiesbn @nickvergessen @DeepDiver1975 @PVince81 @LukasReschke please review.

**Comment by DeepDiver1975** [comment]
This requires heavy upgrade testing by @owncloud/qa

**Comment by DeepDiver1975** [comment]
https://github.com/owncloud/QA/issues/92

**Comment by rullzer** [comment]
@DeepDiver1975 yes it does! 

> owncloud/QA#92

Thanks

**Comment by rullzer** [comment]
Now that the webUI uses the OCS Share API as well. It is time to get this in! So everybody just uses new shares!

CC: @DeepDiver1975 @PVince81 @MorrisJobke @nickvergessen

**Comment by PVince81** [comment]
Had a quick test, set up some reshares on master and also some reshares from inside subfolders as recipient. Ran the upgrade. The shares were flatten properly, looks nice!

**Comment by PVince81** [comment]
Okay, I tested again this time starting on stable8.2, because master isn't suitable for testing because it already creates the shares in the correct format, not legacy.
Worked with link shares, reshares, different files_targets, double shares (shared to same zser through user + group), fed shares, reshared fed shares, reshared subdirs.

**Comment by PVince81** [comment]
Code looks good too :+1:

**Comment by PVince81** [comment]
Needs a seconds reviewer @nickvergessen @MorrisJobke @icewind1991

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Pull request merged**